### PR TITLE
basic sampling

### PR DESCRIPTION
**Pull request by eroell**
**PR Checklist**

-   [x] This comment contains a description of changes (with reason)
-   [x] Referenced issue is linked. Part of #647
-   [x] If you've fixed a bug or added code that should be tested, add tests!
-   [x] Documentation in `docs` is updated

**Description of changes**
New function `ehrapy.pp.sampling(...)` based on [imbalanced-learn](https://imbalanced-learn.org/stable/index.html)

**Technical details**
At the moment, supports `RandomUnderSampler` and `RandomOverSampler`

**Additional context**
Example:

```py
import ehrapy as ep

adata = ep.data.diabetes_130_fairlearn(columns_obs_only=["age"])
print("distribution of age groups:\n", adata.obs.age.value_counts())
adata_balanced = ep.pp.sample(adata, key="age")
print(
    "distribution of age groups after undersampling:\n",
    adata_balanced.obs.age.value_counts(),
)
```

```
distribution of groups:
 age
'Over 60 years'          68541
'30-60 years'            30716
'30 years or younger'     2509

distribution of groups after undersampling:
 age
'30 years or younger'    2509
'30-60 years'            2509
'Over 60 years'          2509

```

**To decide**
- Rather call it sampling, or rather call it balancing?
- Rather make new index for new AnnData, or keep the old, non-unique indices?
- Rather keep all computed fields (.varm, .obsm, etc) or discard them?
- If discard, also discard everything in .var except var name and ehrapy type? There is some things that can end up there by our computations (e.g. highly variable feature statistics), but also user-specific information (e.g. a unit they entered)

### CODE REVIEW

**Comment by Zethson** [comment]
> Rather call it sampling, or rather call it balancing?

I'm probably more in favor of sampling and then we introduce several options and flavors.

> Rather make new index for new AnnData, or keep the old, non-unique indices?

Hmm, non-unique indices sounds like a bad idea

> Rather keep all computed fields (.varm, .obsm, etc) or discard them?

We had this discussion before in a different context and we opted for discard, right?

> If discard, also discard everything in .var except var name and ehrapy type? There is some things that can end up there by our computations (e.g. highly variable feature statistics), but also user-specific information (e.g. a unit they entered)

I'd suggest to keep user-specific information

File: docs/usage/usage.md [hunk 1]
```diff
@@ -74,6 +74,7 @@ Other than tools, preprocessing steps usually don’t return an easily interpret
     :toctree: preprocessing
     :nosignatures:
 
+    preprocessing.sample
```
Comment (Zethson) [inline docs/usage/usage.md:77]: There is now `preprocessing.sample` and `preprocessing.subsample`. https://ehrapy.readthedocs.io/en/latest/usage/preprocessing/ehrapy.preprocessing.subsample.html#ehrapy.preprocessing.subsample is this confusing?

File: ehrapy/preprocessing/_sampling.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+from anndata import AnnData
+from imblearn.over_sampling import RandomOverSampler
+from imblearn.under_sampling import RandomUnderSampler
+
+
+def sample(
+    adata: AnnData,
+    key: str,
+    random_state: int = 0,
+    method: str = "RandomUnderSampler",
```
Comment (Zethson) [inline ehrapy/preprocessing/_sampling.py:17]: Type with Literal

File: ehrapy/preprocessing/_sampling.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        adata: The annotated data matrix of shape `n_obs` × `n_vars`.
+        key: The key in `adata.obs` that contains the group information.
+        random_state: Random seed. Defaults to 0.
+        method: The method to use for balancing. Defaults to "RandomUnderSampler".
+        sampler_kwargs: Keyword arguments for the sample, see the `imbalanced-learn` documentation for options. Defaults to None.
+
+    Returns:
+        A new `AnnData` object, with the balanced groups.
+
+    Example:
```
Comment (Zethson) [inline ehrapy/preprocessing/_sampling.py:35]:
```suggestion
    Examples:
```

File: ehrapy/preprocessing/_sampling.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        >>> import ehrapy as ep
+        >>> adata = ep.data.diabetes_130_fairlearn(columns_obs_only=["age"])
+        >>> print("distribution of age groups:\n", adata.obs.age.value_counts())
+        >>> adata_balanced = ep.pp.sample(adata, key="age")
+        >>> print(
```
Comment (Zethson) [inline ehrapy/preprocessing/_sampling.py:40]: I'd probably remove the prints to keep the examples shorter and more to the point

File: ehrapy/preprocessing/_sampling.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+        >>>  age
+        >>> 'Over 60 years'          68541
+        >>> '30-60 years'            30716
+        >>> '30 years or younger'     2509
+        >>> distribution of groups after undersampling:
+        >>>  age
+        >>> '30 years or younger'    2509
+        >>> '30-60 years'            2509
+        >>> 'Over 60 years'          2509
+
```
Comment (Zethson) [inline ehrapy/preprocessing/_sampling.py:55]:
```suggestion
```

File: ehrapy/preprocessing/_sampling.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    """
+
```
Comment (Zethson) [inline ehrapy/preprocessing/_sampling.py:57]:
```suggestion
```

File: ehrapy/preprocessing/_sampling.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        raise ValueError(f"key not in adata.obs: {key}")
+
+    sampler.fit_resample(adata.X, use_label)
+
+    adata_sampled = adata[sampler.sample_indices_, :].copy()
+
+    # results computed from data should be recomputed if the data changes
+    del adata_sampled.obsm
+    del adata_sampled.varm
+    del adata_sampled.uns
```
Comment (Zethson) [inline ehrapy/preprocessing/_sampling.py:85]: Seems a bit nuclear...

File: tests/preprocessing/test_sampling.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+    # invalid method
+    with pytest.raises(ValueError):
+        ep.pp.sample(adata_mini, key="clinic_day", method="non_existing_method")
+
+    # undersampling
+    adata_sampled = ep.pp.sample(adata_mini, key="clinic_day", method="RandomUnderSampler")
+    assert adata_sampled.n_obs == 4
+    assert adata_sampled.obs.clinic_day.value_counts().min() == adata_sampled.obs.clinic_day.value_counts().max()
+    print(adata_sampled.obs.clinic_day.value_counts())
```
Comment (Zethson) [inline tests/preprocessing/test_sampling.py:34]: prints in tests?

File: tests/preprocessing/test_sampling.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+    # oversampling
+    adata_sampled = ep.pp.sample(adata_mini, key="clinic_day", method="RandomOverSampler")
+    assert adata_sampled.n_obs == 8
+    assert adata_sampled.obs.clinic_day.value_counts().min() == adata_sampled.obs.clinic_day.value_counts().max()
+    print(adata_sampled.obs.clinic_day.value_counts())
```
Comment (Zethson) [inline tests/preprocessing/test_sampling.py:40]: prints in tests?

**Review by Zethson — COMMENTED** [review]
I'm also not confident in the naming and how to handle the other fields. Curious what @Lilly-May thinks (later)

File: tests/preprocessing/test_sampling.py (hunk 8 above)
Comment (eroell) [reply tests/preprocessing/test_sampling.py]: whoops

File: docs/usage/usage.md (hunk 1 above)
Comment (Zethson) [reply docs/usage/usage.md:77]: Yeah, fair argument.

Comment (Lilly-May) [reply docs/usage/usage.md:77]: I agree that having both `ep.pp.sample` and `ep.pp.subsample` is confusing. I think calling it `balancing` is better, but I'm still not 100% convinced. Combining both, i.e., `balanced_sampling`, is too long?

Comment (Zethson) [reply docs/usage/usage.md:77]: We should harmonize the names -> we have `subsample` so maybe `balancedsample` or `balanced_sample`?

**Comment by Lilly-May** [comment]
> Rather keep all computed fields (.varm, .obsm, etc) or discard them?

I would also delete them. Otherwise, it's likely that someone forgets to recalculate things and ends up drawing conclusions from the values calculated for the entire dataset, not the subsampled one.

> If discard, also discard everything in .var except var name and ehrapy type? There is some things that can end up there by our computations (e.g. highly variable feature statistics), but also user-specific information (e.g. a unit they entered)

I also think ideally we would keep the user-specific variables and discard the ones calculated by ehrapy. How would we implement that though? Having a parameter where the user specifies the variables they would like to keep? Doing it the other way around (having a list of vars calculated by ehrapy and deleting those if present) seems like a challenge to maintain...

**Comment by Lilly-May** [comment]
Another thing to consider: The oversampling simply replicates data points, right? Because that will mess up downstream neighbor calculations and thus also UMAP calculations, etc. I don't think there's anything we can do about that except potentially logging a warning for the user?

File: docs/usage/usage.md (hunk 1 above)
Comment (eroell) [reply docs/usage/usage.md:77]: nice inputs, thanks -> both + harmonizing should do

File: ehrapy/preprocessing/_sampling.py (hunk 7 above)
Comment (eroell) [reply ehrapy/preprocessing/_sampling.py:83]: if staying somewhat consistent with scanpy: the closest thing, `sc.pp.subsample`, does not delete any field...
I am somewhat leaning towards that now

Comment (Zethson) [reply ehrapy/preprocessing/_sampling.py]: Then let's do that. Keep it

Comment (Zethson) [reply ehrapy/preprocessing/_sampling.py]: @Lilly-May fine with you? (You can disagree!)

**Comment by Zethson** [comment]
@eroell feel free to merge it after you've resolved the comments above. I prefer API consistency (name + key behavior) with scanpy here over alternatives for now.

**Comment by eroell** [comment]
OK - about duplicated indices in `adata.obs` for the oversampling (or if sampling with replacement):
`sc.pp.subsample` does not mingle with the indices; for somewhat "consistency", I'd suggest we also don't. (the subsampling in scanpy is always without replacement, so never duplicating indices and hence there this consideration never even is necessary)
Calling [reset_index](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.reset_index.html#pandas.DataFrame.reset_index) on the adata.obs if needed could by done by the user after oversampling/sampling with replacement.

**Comment by eroell** [comment]
> @eroell feel free to merge it after you've resolved the comments above. I prefer API consistency (name + key behavior) with scanpy here over alternatives for now.

@Lilly-May keeping the things for scanpy consistency reasons + having the information on duplication in the docs agreeable with you? :)

File: ehrapy/preprocessing/_balanced_sampling.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+
+    Note that `RandomOverSampler <https://imbalanced-learn.org/stable/references/generated/imblearn.over_sampling.RandomOverSampler.html>`_ only replicates observations of the minority groups, which distorts several downstream analyses, very prominently neighborhood calculations and downstream analyses depending on that.
+    The `RandomUnderSampler <https://imbalanced-learn.org/stable/references/generated/imblearn.under_sampling.RandomUnderSampler.html>`_ by  default undersamples the majority group without replacement, not causing this issues of replicated observations.
+
+    Args:
+        adata: The annotated data matrix of shape `n_obs` × `n_vars`.
+        key: The key in `adata.obs` that contains the group information.
+        random_state: Random seed. Defaults to 0.
+        method: The method to use for balancing. Defaults to "RandomUnderSampler".
+        sampler_kwargs: Keyword arguments for the sample, see the `imbalanced-learn` documentation for options. Defaults to None.
```
Comment (Lilly-May) [inline ehrapy/preprocessing/_balanced_sampling.py:34]: Not really important, but do you maybe mean _for the sampler_ or _for sampling_ here?

**Comment by Lilly-May** [comment]
> @Lilly-May keeping the things for scanpy consistency reasons + having the information on duplication in the docs agreeable with you? :)

Looks good to me! I think the way it's implemented now is the most intuitive solution for users👍🏻

File: ehrapy/preprocessing/_balanced_sampling.py (hunk 10 above)
Comment (eroell) [reply ehrapy/preprocessing/_balanced_sampling.py:34]: ah, I see, thanks!

I meant for the sample_r_, I missed the r.

**Pull request merged**